Support mdns when attaching to proxied devices.

chingjun opened the pull request:
Also move the vm service discovery logic into platform-specific implementation of `Device`s. This is to avoid having platform-specific code in attach.dart.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

chingjun committed:
Support mdns when attaching to proxied devices.

Also move the vm service discovery logic into platform-specific
implementation of `Device`s. This is to avoid having platform-specific
code in attach.dart.

christopherfujino commented on packages/flutter_tools/lib/src/proxied_devices/devices.dart:
```diff
@@ -863,3 +893,82 @@ class ProxiedDartDevelopmentService implements DartDevelopmentService {
+          },
+          onError: (Object e) {
+            // Daemon throws string types.
+            if (e is String && e.contains('command not understood')) {
```
Is it safe to ignore other errors here?

christopherfujino commented on packages/flutter_tools/lib/src/proxied_devices/devices.dart:
```diff
@@ -863,3 +893,82 @@ class ProxiedDartDevelopmentService implements DartDevelopmentService {
+          'fuchsiaModule': fuchsiaModule,
+          'filterDevicePort': filterDevicePort,
+          'ipv6': ipv6,
+        }).then(
```
Do we need error handling in case any of the parsing here fails?

christopherfujino replied:
nice, thanks!

chingjun committed:
Add error handling based on reviewer comments

vashworth commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -595,6 +596,33 @@ class FuchsiaDevice extends Device {
+    } on Exception {
+      isolateDiscoveryProtocol?.dispose();
+      final List<ForwardedPort> ports = portForwarder.forwardedPorts.toList();
+      ports.forEach(portForwarder.unforward);
```
So previously, it was
https://github.com/flutter/flutter/blob/137cb4be670d48f3d9b45a5a6c6cd50df488b4a7/packages/flutter_tools/lib/src/commands/attach.dart#L309-L311

Does putting it in a `forEach` wait for each?

christopherfujino reviewed (commented):
This LGTM but I'll defer to Victoria for final approval (she knows this code the best)

chingjun committed:
Updates based on review feedback.

Removed `usesMdns`, and restore platform checking in attach.dart when
showing the message asking user to allow network access on device.

Fix error handling in FuchsiaDevice.

vashworth reviewed (approved):
LGTM!

vashworth commented:
Analyzer issue in test

```
Unused import: '../ios/simulators.dart' • packages/flutter_tools/lib/src/commands/attach.dart:26:8 • unused_import
```

chingjun committed:
Fix analyzer error
